### PR TITLE
Measuring stack usage

### PR DESCRIPTION
**Pull request by noahp**

### CODE REVIEW

**Comment by noahp** [comment]
:white_check_mark: grammarly

File: _drafts/measuring-stack-usage.md [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+     python -c 'import plotext as plt; import sys; \
+       data = [int(x.strip()) for x in sys.stdin.readlines()]; \
+       plt.plot(data, marker="hd"); \
+       plt.title("Stack Usage"); \
+       plt.plot_size(80,25); \
+       plt.theme("pro"); \
+       plt.show()'
+
+                                     Stack Usage
+     ┌─────────────────────────────────────────────────────────────────────────┐
```
Comment (noahp) [inline _drafts/measuring-stack-usage.md:300]: this renders so badly on mobile SIGH. replace with png :(.
image

Comment (noahp) [reply _drafts/measuring-stack-usage.md]: sigh switched to gnuplot too. RIP my one-liner.

File: _drafts/measuring-stack-usage.md [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+>
+> ![](/img/stacks/stack-growth-double-link.excalidraw.svg)
+>
+> Read more about that here:
+>
+> <https://blog.japaric.io/stack-overflow-protection/>
+
+### RTOS stacks
+
+![](/img/stacks/rtos-stacks.excalidraw.svg)
```
Comment (noahp) [inline _drafts/measuring-stack-usage.md:71]: Fix diagram, task->thread

File: _drafts/measuring-stack-usage.md [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+variable will then access the stack-allocated 32-bit region.
+
+In a typical single-threaded embedded program that uses both stack and heap
+(`malloc` and friends), the memory ordering might look like this:
+
+![](/img/stacks/stack-growth.excalidraw.svg)
+
+You can find some more explanation here:
+
+<https://courses.engr.illinois.edu/cs225/fa2022/resources/stack-heap/>
```
Comment (ejohnso49) [inline _drafts/measuring-stack-usage.md:57]: ❤️ for my alma mater

File: _drafts/measuring-stack-usage.md [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+  while (stack < stack_end && *stack == STACK_FILL_VALUE) {
+    stack++;
+  }
+
+  return (uintptr_t)stack;
+}
+```
+
+This strategy can be used after the system has run significant pieces of its
+functionality, to give a reasonably accurate measurement of the maximum stack
```
Comment (ejohnso49) [inline _drafts/measuring-stack-usage.md:202]:
```suggestion
Use this strategy after the system executes core functionality to collect a reasonably
accurate measurement of the maximum stack usage.
```

File: _drafts/measuring-stack-usage.md [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+usage.
+
+_Note: since this approach relies on values being written to the stack, if a
+buffer is allocated on stack but only partially used, the full buffer size
```
Comment (ejohnso49) [inline _drafts/measuring-stack-usage.md:206]:
```suggestion
buffer is allocated on the stack but only partially used, the full buffer size
```

File: _drafts/measuring-stack-usage.md [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+wouldn't be immediately represented in the fill pattern._
```
Comment (ejohnso49) [inline _drafts/measuring-stack-usage.md:207]:
```suggestion
will not be counted as used. Any memory still containing the fill pattern will be counted as unused_
```

File: _drafts/measuring-stack-usage.md [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+        name = gdb.selected_frame().name()
+        # if we just changed function name, record the high water mark
+        # from the last function
+        if current_function is not None and name != current_function:
+            f.write(f"{current_function}|{current_function_max_su}\n")
+            current_function_max_su = 0
+        current_function = name
+        if su > current_function_max_su:
+            current_function_max_su = su
+        # f.write(f"{current_function}|{current_function_max_su}\n")
```
Comment (ejohnso49) [inline _drafts/measuring-stack-usage.md:274]: Should this be uncommented?

File: img/stacks/stack-painting.excalidraw.png
Comment (ejohnso49) [inline img/stacks/stack-painting.excalidraw.png:1]: I think adding the addresses at the top and bottom of the blocks like you did in the other diagrams would be nice here

**Review by ejohnso49 — COMMENTED** [review]
Love it

File: _drafts/measuring-stack-usage.md (hunk 7 above)
Comment (noahp) [reply _drafts/measuring-stack-usage.md:274]: ah that was for debug, good spot! deleting it

**Review by tyhoff — APPROVED** [review]

**Pull request merged**